R/zchunk_LA1011.en_bal_adj.R

```diff
@@ -35,54 +36,174 @@ module_energy_LA1011.en_bal_adj_DISABLED <- function(command, ...) {
+
+    # This is complicated. Because the output of "gas works gas" that is produced from coal is not distinguished from that produced by other fuels,
+    # the process is modeled based on the fuel inputs, and the output fuel is assigned the same name as natural gas. As a result the estimates of
+    # TPES at this point for natural gas include both natural gas and gasified coal. This subtraction generally follows the method used in code file L122.
+    # Heat production from district heat sector
+    A22.globaltech_coef %>%
+      gather(year, value, -supplysector, -subsector, -technology, -minicam.energy.input) %>%
+      mutate(year = as.integer(year))  %>%
+      # Adding empty historical years to fill in with interpolation
+      complete(year = unique(c(HISTORICAL_YEARS, year)),
+               nesting(supplysector, subsector, technology, minicam.energy.input)) %>%
+      arrange(year) %>%
+      group_by(technology, subsector, supplysector, minicam.energy.input)  %>%
+      # Interpolate to fill in missing globaltech_coef historical years
+      mutate(value = approx_fun(year, value)) %>%
+      left_join(distinct(calibrated_techs), by = c("supplysector", "subsector", "technology", "minicam.energy.input")) %>%
+      select(supplysector, subsector, technology, minicam.energy.input, year, value, sector, fuel) %>%
+      filter(supplysector == "gas processing") -> L1011.gasproc_coef
+
+    L1011.en_bal_EJ_R_Si_Fi_Yh %>%
+      filter(sector == "in_gas processing" & fuel=="coal") %>%
+      mutate(sector.x = sub("in_","", sector)) %>%
+      left_join(L1011.gasproc_coef %>%
+                  rename(sector.x = sector), by = c("sector.x", "fuel", "year")) %>%
+      mutate(value = value.x / value.y) %>%
+      select(GCAM_region_ID, sector, fuel, year, value) %>%
+      # Changing the fuel to gas, sector to TPES so we can easily join it with gas in the next step
+      mutate(fuel = "gas") %>%
+      mutate(sector = "TPES") -> L1011.out_EJ_R_gasproc_coal_Yh
+
+    # Subtract
+    L1011.en_bal_EJ_R_Si_Fi_Yh %>%
+      left_join(L1011.out_EJ_R_gasproc_coal_Yh, by = c("GCAM_region_ID", "sector", "fuel", "year"))%>%
+      mutate(value.x = if_else(is.na(value.y), value.x, value.x-value.y)) %>%
+      rename(value = value.x) %>%
+      select(GCAM_region_ID, sector, fuel, year, value)-> L1011.en_bal_EJ_R_Si_Fi_Yh
+
+    # This is also complicated. In regions with very low natural gas use and high coal-to-gas with very high input-output coefs on coal-to-gas production
+    # (South Africa), dividing the coal input by the IO coef may cause gas production in excess of the demands in the region.
+    # If this is the case, need to return to original energy balance data and reduce the coal input to gas works (can re-allocate to another sector if desired)
+
+    if( any( L1011.en_bal_EJ_R_Si_Fi_Yh[ L1011.en_bal_EJ_R_Si_Fi_Yh$sector == "TPES" & L1011.en_bal_EJ_R_Si_Fi_Yh$fuel == "gas", X_historical_years ] < 0 ) ){
```

Comment:
Errors. Please *do not* open a PR for non-running code. This wastes other people's time.